core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -301,6 +303,119 @@ class TransactionStateManager(brokerId: Int,
+      if (responseStatus.size != 1 || !responseStatus.contains(topicPartition))
+        throw new IllegalStateException("Append status %s should only have one partition %s"
+          .format(responseStatus, topicPartition))
+
+      val status = responseStatus(topicPartition)
+
+      var responseError = if (status.error == Errors.NONE) {
+        Errors.NONE
+      } else {
+        debug(s"Transaction state update $txnMetadata for $transactionalId failed when appending to log " +
+          s"due to ${status.error.exceptionName}")
+
+        // transform the log append error code to the corresponding coordinator error code
+        status.error match {
+          case Errors.UNKNOWN_TOPIC_OR_PARTITION
+               | Errors.NOT_ENOUGH_REPLICAS
+               | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND
+               | Errors.REQUEST_TIMED_OUT => // note that for timed out request we return NOT_AVAILABLE error code to let client retry
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.COORDINATOR_NOT_AVAILABLE} to the client")
+
+            Errors.COORDINATOR_NOT_AVAILABLE
+
+          case Errors.NOT_LEADER_FOR_PARTITION =>
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.NOT_COORDINATOR} to the client")
+
+            Errors.NOT_COORDINATOR
+
+          case Errors.MESSAGE_TOO_LARGE
+               | Errors.RECORD_LIST_TOO_LARGE
+               | Errors.INVALID_FETCH_SIZE =>
+
+            error(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning UNKNOWN error code to the client")
+
+            Errors.UNKNOWN
+
+          case other =>
+            error(s"Appending metadata message f$txnMetadata or $transactionalId failed due to " +
+              s"unexpected error: ${status.error.exceptionName}")
+
+            other
+        }
+      }
+
+      // now try to update the cache: we need to update the status in-place instead of
+      // overwriting the whole object to ensure synchronization
+      getTransaction(transactionalId) match {
+        case Some(metadata) =>
+          metadata synchronized {
+            if (responseError == Errors.NONE &&
+              metadata.pid == txnMetadata.pid &&
+              metadata.epoch == txnMetadata.epoch &&
+              metadata.txnTimeoutMs == txnMetadata.txnTimeoutMs &&
+              metadata.completeTransitionTo(txnMetadata.state)) {
+              // only topic-partition lists could possibly change (state should have transited in the above condition)
+              metadata.addPartitions(txnMetadata.topicPartitions.toSet)
+            } else {
+              // the record has already been appended but the cache has been modified.
```

Comment:
One possibility I had in mind is that producer-1 is stalled while waiting for the appended txn entry to be replicated, and producer-2 grabbed this transactional id and caused the old txn to abort and probably completed another txn; then when this replication finally get returned it will see this.

But I may change my mind when implementing the completion during initPid, so this logic could change.

I will keep this in mind and we can further discuss about this.